service/controller/resource/cleanupendpointips/create.go

```diff
@@ -149,28 +154,54 @@ func removeFromEndpointAddressList(addresses []corev1.EndpointAddress, indexesTo
 	return newAddresses
 }
 
+func controlPlanePodForTCNode(node corev1.Node, pods []corev1.Pod) (corev1.Pod, error) {
+	for _, pod := range pods {
+		if pod.Name == node.Name {
+			return pod, nil
+		}
+	}
+	// Unless there is a race condition where the Pods are modified
+	// after being checked by podsEqualNodes(), this should never be reached.
+	return corev1.Pod{}, microerror.Maskf(noPodForNodeError, fmt.Sprintf("no control plane pod for tenant cluster node %s", node.Name))
+}
+
 // removeDeadIPFromEndpoints compares endpoint IPs with current state of nodes and
 // removes any IP addresses that does not belong to any node.
-func removeDeadIPFromEndpoints(endpoints *corev1.Endpoints, nodes []corev1.Node) (int, *corev1.Endpoints) {
+func removeDeadIPFromEndpoints(endpoints *corev1.Endpoints, nodes []corev1.Node, cpPods []corev1.Pod) (int, *corev1.Endpoints, error) {
 	endpointAddresses := endpoints.Subsets[0].Addresses
 
 	var indexesToDelete []int
 	for i, ip := range endpointAddresses {
 		found := false
 		// check if the ip belongs to any k8s node
 		for _, node := range nodes {
+
 			if node.Labels["ip"] == ip.IP {
-				found = true
-				break
+				// Find the control plane pod representing this node
+				cpPod, err := controlPlanePodForTCNode(node, cpPods)
+				if err != nil {
+					return len(indexesToDelete), endpoints, microerror.Mask(err)
+				}
+
+				// Check if the CP pod is Ready
+				for _, c := range cpPod.Status.Conditions {
```

Comment:
It can make the code cleaner to move this into a `key` function. Then you could call it like `if key.PodIsReady(cpPod)`.